pageserver/src/relish_storage/storage_sync.rs

```diff
@@ -0,0 +1,1416 @@
+//! A synchronization logic for the [`RelishStorage`] and the state to ensure the correct synchronizations.
+//!
+//! The synchronization does not aim to be immediate, instead
+//! doing all the job in a separate thread asynchronously, attempting to fully replicate the
+//! pageserver timeline workdir data on the remote storage.
+//!
+//! [`SYNC_QUEUE`] is a priority queue to hold [`SyncTask`] for image upload/download.
+//! The queue gets emptied by a single thread with the loop, that polls the tasks one by one.
+//!
+//! During the loop startup, an initial loop state is constructed from all remote storage entries.
+//! It's enough to poll the remote state once on startup only, due to agreement that the pageserver has
+//! an exclusive write access to the relish storage: new files appear in the storage only after the same
+//! pageserver writes them.
+//!
+//! The list construction is currently the only place where the storage sync can return an [`Err`] to the user.
+//! New upload tasks are accepted via [`schedule_timeline_upload`] function disregarding of the corresponding loop startup,
+//! it's up to the caller to avoid uploading of the new relishes, if that caller did not enable the loop.
+//! After the initial state is loaded into memory and the loop starts, any further [`Err`] results do not stop the loop, but rather
+//! reschedules the same task, with possibly less files to sync in it.
+//!
+//! The synchronization unit is an image: a set of layer files (or relishes) and a special metadata file.
+//! Both upload and download tasks consider image in a similar way ([`LocalTimeline`] and [`RemoteTimeline`]):
+//! * a set of relishes (both upload and download tasks store the files as local pageserver paths, ergo [`PathBuf`] is used).
+//! * a set of ids to distinguish the images ([`ZTenantId`] and [`ZTimelineId`])
+//! * `disk_consistent_lsn` which indicates the last [`Lsn`] applicable to the data stored in this image.
+//!
+//! The same relish has identical layer paths in both structs, since both represent the relish path in pageserver's workdir.
+//! This way, the sync can compare remote and local images seamlessly, downloading/uploading missing files if needed.
+//!
+//! `disk_consistent_lsn` and the list of files written gets stored in the metadata file of every image after a successful image checkpoint and
+//! it's always required for an upload: getting the metadata correctly is the main reason why image uploads are done after checkpointing.
+//! Timelines with no such [`Lsn`] cannot guarantee their local file consistency and are not considered for backups.
+//! Not every upload of the same timeline gets processed: if `disk_consistent_lsn` is unchanged, the remote timeline is not updated.
+//!
+//! Remote timelines may lack `disk_consistent_lsn` if their metadata file is corrupt or missing.
```

Comment:
> It will successfully grab the newer metadata file from disk.

That seems like an error to me, very good catch: I've forgotten that metadata always updates.
I can think of passing the whole metadata to `schedule_timeline_upload` instead of just `disk_consistent_lsn` and then serialize it to bytes and upload in the end.

The other way could be to use different metadatas, named differently (as we've discussed once), but I'm still not sure this isn't the way that brings more issues than it solves.